# 2 - Clasificación

**Sumario**

1. Introducción
2. Funciones de pérdida
3. Métricas
4. Clasificación binaria
5. Clasificación multiclase
6. Clasificación multietiqueta
7. Optimización de hiperparámetros

## 2.1 - Introducción

En tareas de clasificación, nuestro objetivo es construir un modelo capaz de **predecir correctamente una o varias etiquetas para cada uno de las instancias de entrada**. 

En la siguiente figura se presenta uno de los problemas más comunes de clasificación: se quiere construir un modelo capaz de clasificar un conjunto de *emails* cuya clase está formada por dos etiquetas: 
1. correo válido
2. correo spam

<img src="images_2/email_classification.png" width="500" data-align="center">

El tipo de comportamiento del modelo vendrá definido por la estructura de la clase mediante la que están definidos los diferentes ejemplos. En función del tipo de clase que hayamos definido, podemos diferenciar tres tipos de modelos:
* **Clasificación binaria** (una etiqueta, dos clases).
* **Clasificación multiclase** (una etiqueta, múltiples clases).
* **Clasificación multietiqueta** (múltiples etiquetas, dos clases).

## 2.2 - Funciones de pérdida (*loss*)

A la hora de efectuar un proceso de clasficación, podemos aplicar diferentes funciones de pérdida dependiendo del tipo de salida que queramos obtener. Asi pues:
* **Clasificación binaria**
    * Binary crossentropy loss
* **Clasificación multiclase**
    * Multiclass crossentropy loss
    * Sparse multiclass crossentropy loss
* **Clasificación multietiqueta**
    * Binary crossentropy loss

### 2.2.1 - Binary crossentropy

La **entropía cruzada binaria** (binary crossentropy) es un tipo de función de pérdida utilizada para la construcción de modelos de **clasificación binaria**. Se calcula mediante la siguiente fórmula:

$$
f(y_{i}, \hat{y}_{i}) = -\frac{1}{n} \sum_{i=1}^{n} y_{i} \log(\hat{y}_{i}) + (1 - y_{i}) \log(1 - \hat{y}_{i})
$$

donde
* $n$ es el número de instancias de entrenamiento utilizadas para calcular el valor de pérdida.
* $y_{i}$ es el valor de salida esperado
* $\hat{y}_{i}$ es el valor de salida real

Para poder utilizar este tipo de función de pérdida es necesario recurrir a la **función sigmoidea como función de activación de la última capa de la red**, ya que es la única compatible con esta función de pérdida. Esto se debe a que la función de pérdida debe calcular el logaritmo de $y_{i}$, que solo existe cuando el valor de $\hat{y}_{i}$ se sitúa entre $0$
y $1$.

<img src="images_2/binary_crossentropy.png" width="700" data-align="center">

**Nota:** La función softmax (con 2 valores) también valdría en este caso ya que al fin y al cabo **la función sigmoidea no es más que una versión binaria de la función softmax**.

**Extra:** [**Derivación de la función de pérdida para su aplicación con el descenso por gradiente**](https://www.python-unleashed.com/post/derivation-of-the-binary-cross-entropy-loss-gradient)

### 2.2.2 - Multiclass crossentropy

La **entropía cruzada multiclase o categórica**  (multiclass crossentropy or categorical crossentropy) es un tipo de función de pérdida utilizada para la construcción de modelos de **clasificación multiclase**. Se calcula mediante la siguiente fórmula:

$$
f(y_{i}, \hat{y}_{i}) = - \sum_{i}^{n} y_{i} \log \hat{y}_{i}
$$

donde
* $n$ es el número de instancias de entrenamiento utilizadas para calcular el valor de pérdida.
* $y_{i}$ es el valor de salida esperado
* $\hat{y}_{i}$ es el valor de salida real

Para poder utilizar este tipo de función de pérdida, se recomienda usar la **función softmax como función de activación de la última capa de la red**, pues esta solo **necesita que la salida del modelo sea positiva** (por el logaritmo). Por tanto, la función softmax se adapta perfectamente, ya que efectúa una reescalada de la salida, de manera que todos los valores son expresados entre $0$ y $1$.

<img src="images_2/multiclass_crossentropy.png" width="700" data-align="center">

**Extra:** [**Derivación de la función de multiclass crossentropy loss para su aplicación con el descenso por gradiente**](https://towardsdatascience.com/derivative-of-the-softmax-function-and-the-categorical-cross-entropy-loss-ffceefc081d1)

### 2.2.3 - Sparse multiclass crossentropy

La **entropía cruzada multiclase dispersa** (sparse multiclass crossentropy) es un tipo de función de pérdida utilizada para la construcción de modelos de **clasificación multiclase cuando el número de clases es muy elevado**. 

Existen situaciones donde el número posible de clases es muy grande,. or ejemplo, si tenemos que clasificar marcas a partir de una descripción de producto (hay millones de marcas en el mundo). Esta función de pérdida **realiza el mismo cálculo que la función de entropía cruzada multiclase**, pero **sin necesidad de que el vector de entrada sea una codificación de tipo one-hot**. Por tanto, se calcula mediante la siguiente fórmula:

$$
f(y_{i}, \hat{y}_{i}) = - \sum_{i}^{n} y_{i} \log \hat{y}_{i}
$$


<img src="images_2/sparse_multiclass_crossentropy.png" width="700" data-align="center">

## 2.3 - Métricas

Para evaluar los modelos de clasificación, existen diferentes tipos de métricas de "bondad".

La **matriz de confusión** (confusion matrix) es un procedimiento para extraer la información necesaria y calcular las diferentes métricas de "bondad" utilizadas por los algoritmos de clasificación. Consiste en una matriz $n \times n$, donde $n$ es el número de clases, que describe el rendimiento del modelo a partir de los datos del conjunto de
test. Su denominación alude a su finalidad: **identificar dónde está confundiendo las clases el modelo**.

En el caso más básico, se aplica a una clasificación binaria una matriz de confusión formada por cuatro características:
* **Verdaderos positivos (VP)**. Aquellos ejemplos del conjunto de test cuya clase esperada es 1 (verdadero) y el resultado generado por el modelo de aprendizaje es 1 (verdadero).
* **Verdaderos negativos (VN)**. Aquellos ejemplos del conjunto de test cuya clase esperada es 0 (negativo) y el resultado generado por el modelo de aprendizaje es 0 (negativo).
* **Falsos positivos (FP)**. Aquellos ejemplos del conjunto de test cuya clase esperada es 0 (falso) y el resultado generado por el modelo de aprendizaje es 1 (verdadero).
* **Falsos negativos (FN)**. Aquellos ejemplos del conjunto de test cuya clase esperada es 1 (Verdadero) y el resultado generado por el modelo de aprendizaje es 0 (Falso).

A continuación, vamos a definir las diferentes métricas de bondad que se obtienen a través de la matriz de confusión a partir del siguiente ejemplo:

<table>
    <tr>
        <th>Matriz de confusión binaria</th>
        <th>Ejemplo</th>
    </tr>
    <tr>
        <td><img src="images_2/matriz_confusion_1.png" width="500" data-align="center"></td>
        <td><img src="images_2/matriz_confusion_3.png" width="500" data-align="center"></td>
    </tr>
</table>

**Nota:** Esta representación también se puede aplicar a modelos de clasificación no binaria como, por ejemplo, aquellos que presentan tres clases:

<img src="images_2/matriz_confusion_2.png" width="500" data-align="center">

### 2.3.1 - Accuracy

La **exactitud** (**accuracy** en inglés) se define como **el grado de concordancia entre el resultado generado por el modelo y el resultado real**. Se obtiene a través de la siguiente fórmula:

$$
A = \frac{VN + VP}{VP + VN +FP + FN}
$$

Con respecto a los datos del ejemplo definido anteriormente, obtendríamos un accuracy de $0.92$:

$$
A = \frac{87 + 5}{87 + 5 + 2 + 6} = 0.92
$$

El accuracy **no es una métrica adecuada para situaciones donde las clases (labels) del conjunto de datos se encuentran desbalanceadas**. En estos casos, resulta mucho más conveniente usar las métricas de precision, recall o F1-Score.

### 2.3.2 - Precision

La **precisión** (**precision** en ingles) es una medida centrada en el modelo porque nos indica **el grado de precision del modelo cuando predice una determinada clase**. Por ejemplo, en el caso de las imagenes de animales, cuando nuestro modelo que una imagen es un perro, cual es la probabilidad de que sea esto cierto. Se obtiene a través de la siguiente fórmula (ejemplo para 2 clases, centrándonos en la clase "positivo"):

$$
\text{Precision} = \frac{VP}{VP + FP}
$$

Con respecto a los datos del ejemplo definido anteriormente, obtendríamos una precision de $0.45$:

$$
\text{Precision} = \frac{5}{5 + 6} = 0.45
$$

### 2.3.3 - Recall

La **exhaustividad** (**recall** en inglés) es una medida centrada en los datos porque nos indica **como de bien se predice una clase de los datos**. Por ejemplo, en el caso de las imagenes de animales, cómo de bien identifica nuestro modelo a los perros. Se obtiene a través de la siguiente fórmula ejemplo para 2 clases, centrándonos en la clase "positivo"):

$$
\text{Recall} = \frac{VP}{VP + FN}
$$

Con respecto a los datos del ejemplo definido anteriormente, obtendríamos una precision de $0.62$:

$$
\text{Recall} = \frac{5}{2 + 6} = 0.62
$$

### 2.3.4 - F1 score

La **puntuación F1** (**F1-score**) es la **media harmónica de la precisón y de la exhaustividad (recall)**. Así esta métrica es útil ya que nos permite combinar ambas métricas:

$$
\text{F1-score} = 2 * \frac{P * R}{P + R}
$$

Con respecto a los datos del ejemplo definido anteriormente, obtendríamos una F1-score de $0.52$:

$$
\text{F1-score} = 2 * \frac{0.45 * 0.62}{0.45 + 0.62} = 0.52
$$

### 2.3.5 - F$\beta$ score

La puntuación F$\beta$ (**F$\beta$ score**) es una métrica que **combina la precisión y la exhaustividad (recall) con un factor de variación** $\beta$ que permite modificar el grado de importancia que deseamos asigna a cada una de estas métricas. **Cuanto más elevado el valor de $\beta$, mayor es la importancia que concedemos al *recall***:

$$
\text{F1-score} = (1 + \beta^{2}) * \frac{P * R}{(\beta * P) + R}
$$

## 2.4 - Clasificación binaria

A continuación vamos a describir cómo construir una red neuronal para un problema de clasificación binaria.

### 2.4.1 - Preparación de datos

La primera fase de todo proceso de ML es la preparación de los datos. En este caso, vamos a utilizar **el conjunto de datos sobre pruebas diagnósticas para la detección del cáncer de mama**.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf

data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data", header=None)
data.columns = [
    "Sample_code_number",
    "Clump_Thickness",
    "Uniformity_of_Cell_Size",
    "Uniformity_of_Cell_Shape",
    "Marginal_Adhesion",
    "Single_Epithelial_Cell_Size",
    "Bare_Nuclei",
    "Bland_Chromatin",
    "Normal_Nucleoli",
    "Mitoses",
    "Class"
]

Una vez realizada la carga de los datos, debemos llevar a cabo una serie de transformaciones y modificaciones en la información del conjunto de datos con el objetivo de adecuarlos y poder construir nuestra red de neuronas:

In [2]:
# Eliminación de la columna que contiene el id de las muestras 
data.drop("Sample_code_number", axis="columns", inplace=True)

# Modificación de los valores ? que existen en la columna Bare Nuclei (en este caso por el valor -1)
data["Bare_Nuclei"] = data['Bare_Nuclei'].replace('?', '-1')

# Modificación de las etiquetas para que los valores sean 0 y 1 
data["Class"] = data["Class"].map(lambda x: 1 if x == 4 else 0)

# Generación de los conjuntos de entrenamiento y test split_handler (en este caso no vamos a generar de validación aqui, lo haremos mas adelante)
# split proportion: 0.88 train, 0.12 test
train_set, test_set = train_test_split(data, test_size=0.12, random_state=0, stratify=data["Class"])

# Transformación del conjunto de entrenamiento en tensores 
X_train = tf.convert_to_tensor (train_set.iloc[:, :9], np.int32) 
y_train = tf.convert_to_tensor (train_set.iloc[:, 9:], np.int8)

# Transformación del conjunto de test en tensores 
X_test = tf.convert_to_tensor (test_set.iloc[:, :9], np.float32)
y_test = tf.convert_to_tensor (test_set.iloc[:, 9:], np.int8)

**Transformamos los conjuntos de datos en tensores porque ese es el formato esperado por la red para los datos**

### 2.4.2 - Construcción de la red

Una vez preparados los datos, podemos construir nuestra red de neuronas. Para ello, vamos a crear una red formada por tres capas mediante Keras:

* Una capa de entrada que acepte tensores unidimensionales de tamaño 9. Para ello vamos a usar `keras.layers.Flatten`.
* Una capa densa con 18 neuronas cuya función de activación sea una ReLU.
* Una capa densa con una sola neurona con una función de activación de tipo sigmoidea, de manera que nos devuelva un valor comprendido entre 0 y 1 que se corresponderá con una de las dos clases que podemos obtener.

In [3]:
from tensorflow import keras

# Creación de las capas que forman la estructura de la red
layers = [keras.layers.Flatten(input_shape=(9,)),
          keras.layers.Dense(18, activation=tf.nn.relu),
          keras.layers.Dense(1, activation=tf.nn.sigmoid),
         ]

# Compilación de la red de neuronas, agrupando las diferentas capas de forma secuencial
model = keras.Sequential(layers, name="binary_classification_model")

# Configuración del algoritmo de optimización y de la función de loss
model.compile(optimizer="sgd",
              loss="binary_crossentropy",
              metrics=["accuracy"]
             )

Tras la compilación, podemos comprobar la estructura de nuestra red mediante la función `summary`

In [4]:
model.summary()

Model: "binary_classification_model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 9)                 0         
                                                                 
 dense (Dense)               (None, 18)                180       
                                                                 
 dense_1 (Dense)             (None, 1)                 19        
                                                                 
Total params: 199
Trainable params: 199
Non-trainable params: 0
_________________________________________________________________


Con respecto a los parámetros observamos lo siguiente:
* La capa de entrada `flatten_1` no tiene parámetros ya que simplemente nos sirve para introducir información en la red.
* La capa densa `dense_2` está formada por 18 neuronas, cada una de ellas con 9 parámetros de entrada y un parámetro de *bias*, haciendo un total de **180**.
* La capa densa `dense_3` está formada por 1 neurona con 18 parámetros de entrada y un parámetro de *bias*, haciendo un total de **19**.

### 2.4.3 - Entrenamiento

Una vez compilada nuestra red de neuronas, podemos abordar el proceso de entrenamiento mediante la función `fit`. En este caso, hemos seleccionado las siguientes opciones:
* Número de iteraciones (*epochs*): 25
* Tamaño del batch de entrenamiento: 100 ejemplos
* Tamaño del conjunto de validación: 8% del conjunto de entrenamiento

In [5]:
import datetime
from time import time
from keras.callbacks import TensorBoard

# Definición de los callback de TF Board
tensorboard_callback = TensorBoard(log_dir="logs")

# Ejecución del proceso de aprendizaje
model.fit(
    X_train,
    y_train,
    epochs=25, # Numero de iteraciones
    batch_size=100, # Tamaño de los batches
    validation_split=0.08, # Tamaño del conjunto de validación
    callbacks=[tensorboard_callback]
)

Epoch 1/25
6/6 [==============================] - 1s 68ms/step - loss: 0.7494 - accuracy: 0.6159 - val_loss: 0.6707 - val_accuracy: 0.6800
Epoch 2/25
6/6 [==============================] - 0s 12ms/step - loss: 0.6442 - accuracy: 0.6159 - val_loss: 0.6823 - val_accuracy: 0.6400
Epoch 3/25
6/6 [==============================] - 0s 13ms/step - loss: 0.6191 - accuracy: 0.6779 - val_loss: 0.6642 - val_accuracy: 0.6800
Epoch 4/25
6/6 [==============================] - 0s 12ms/step - loss: 0.5875 - accuracy: 0.7274 - val_loss: 0.6292 - val_accuracy: 0.8200
Epoch 5/25
6/6 [==============================] - 0s 10ms/step - loss: 0.5621 - accuracy: 0.8248 - val_loss: 0.6340 - val_accuracy: 0.8000
Epoch 6/25
6/6 [==============================] - 0s 13ms/step - loss: 0.5430 - accuracy: 0.8442 - val_loss: 0.6245 - val_accuracy: 0.8200
Epoch 7/25
6/6 [==============================] - 0s 13ms/step - loss: 0.5242 - accuracy: 0.8407 - val_loss: 0.5810 - val_accuracy: 0.8600
Epoch 8/25
6/6 [===========

In [6]:
# Evaluación del modelo mediante el conjunto de test
test_loss, test_acc = model.evaluate(X_test, y_test)

3/3 [==============================] - 0s 3ms/step - loss: 0.3468 - accuracy: 0.9167


### 2.4.4 - Inferencia y visualización

Una vez finalizado el proceso de aprendizaje, podemos recurrir a TensorBoard para comprobar la evolución del proceso de entrenamiento.

**Nota:** Dado el rapido aprendizaje del modelo, TensorBoard se confunde un poco al generar los gráficos ya que está acostumbrado a que se siga un proceso con forma "logarítmica". [Para ver un ejemplo más "común", podemos ejecutar el siguiente notebook en Google Colab.](https://colab.research.google.com/github/tensorflow/tensorboard/blob/master/docs/tensorboard_in_notebooks.ipynb#scrollTo=ixZlmtWhMyr4)


#### Ejecución en local

En caso de que estemos ejecutando el notebook en nuestra máquina local, simplemente tendríamos que situarnos con la terminal en el directorio del notebook y correr el siguiente comando:
```
tensorboard --logdir logs
```
Se nos indicará entonces cual es la dirección local a la que tenemos que acceder y mostrará la aplicación de TensorBoard con los logs de la ejecución actual:

<img src="images_2/tensorboard.png" width="800" data-align="center">

#### Ejecución en Google Colab

En caso de que estemos ejecutando este notebook en Google Colab, no podriamos acceder al directorio local y en su lugar deberiamos utilizar una extensiónde que nos permitiría generar TensorBoard dentro de Jupyter

In [7]:
# Solo en Google Colab
# %load_ext tensorboard

In [8]:
# %tensorboard --logdir logs

## 2.5 - Clasificación multiclase

A continuación vamos a describir cómo construir una red neuronal para un problema de clasificación multiclase.

### 2.5.1 - Preparación de datos

Como ya sabemos, la primera fase de todo proceso de ML es la preparación de los datos. En este caso, **vamos a utilizar el famoso dataset Iris**, que consiste en clasificar flores en tres variedades según carateristicas del pétalo y el sépalo de la flor.

In [9]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf

data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data", header=None)
data.columns = [
    "sepal_length",
    "sepal_width",
    "petal_length",
    "petal_width",
    "class"
]

Una vez realizada la carga de los datos, debemos llevar a cabo una serie de transformaciones y modificaciones en la información del conjunto de datos con el objetivo de adecuarlos y poder construir nuestra red de neuronas:
* Codificación de las clases en formato de numero entero (índice).
* Codificación de la clases en formato *one-hot*.
* Codificación numérica de todos los atributos en formato decimal (*float*).
* Generación de los splits de entrenamiento y test.
* Transformación de los splits a formato *tensor* para poder ser ingeridos por la red.

In [10]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Definimos un codificador para transformar las clases en números enteros
class_encoder = LabelEncoder()
class_encoder.fit (data["class"])
integer_class_representation = class_encoder.transform (data["class"])

# Definimos un codificador para transformar las clases de formato entero a formato one-hot
one_hot_encoder = OneHotEncoder()

# Generación de los conjuntos de entrenamiento y test split_handler (en este caso no vamos a generar de validación aqui, lo haremos mas adelante)
# split proportion: 0.88 train, 0.12 test
train_set, test_set = train_test_split(data, test_size=0.12, random_state=0, stratify=data["class"])

# Transformación del conjunto de entrenamiento en tensores 
X_train = tf.convert_to_tensor(train_set.iloc[:, :4], np.int32)
y_train = tf.convert_to_tensor(one_hot_encoder.fit_transform(train_set.iloc[:, 4:]).toarray())

# Transformación del conjunto de test en tensores 
X_test = tf.convert_to_tensor(test_set.iloc[:, :4], np.int32)
y_test = tf.convert_to_tensor(one_hot_encoder.fit_transform(test_set.iloc[:, 4:]).toarray())

En vez de utilizar el one-hot encoder, podemos utilizar la forma entera de las clases con la función de pérdida 

In [11]:
# Transformación del conjunto de entrenamiento en tensores donde la clase está en formato entero, en vez de en formato one-hot
y_train_integer = tf.convert_to_tensor(class_encoder.fit_transform(train_set.iloc[:, 4:])) 

# Transformación del conjunto de test en tensores donde la clase está en formato entero, en vez de en formato one-hot
y_test_integer = tf.convert_to_tensor(class_encoder.fit_transform(test_set.iloc[:, 4:]))

C:\Users\rofe2001\Anaconda3\envs\nlp\lib\site-packages\sklearn\preprocessing\_label.py:115: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


### 2.5.2 - Construcción de la red

Una vez preparados los datos, podemos construir nuestra red. En este sentido, vamos a crear una red formada por tres capas mediante Keras:
* Una capa de de tipo `flatten` que acepte como entrada tensores unidimensionales de dimensión 4.
* Una capa densa de 8 neuronas con una función de activación de tipo ReLU.
* Una capa densa de 3 neuronas con una función de activación de tipo `softmax` que nos devolverá la probabilidad de que el ejemplo pertenezca a cada una de las clases.

In [12]:
from tensorflow import keras

# Creación de las capas que forman la estructura de la red
layers_onehot = [keras.layers.Flatten(input_shape=(4,)),
                 keras.layers.Dense(8, activation=tf.nn.relu),
                 keras.layers.Dense(3, activation=tf.nn.softmax),
                ]

# Compilación de la red de neuronas, agrupando las diferentas capas de forma secuencial
model_onehot = keras.Sequential(layers_onehot, name="multiclass_classification_model_onehot")

# Configuración del algoritmo de optimización y de la función de loss
# Nota, en este caso vamos a usar Adam en vez de SGD simplemente para demostrar 
# como se le llama en Keras
model_onehot.compile(optimizer="adam",
              loss="categorical_crossentropy",
              metrics=["accuracy"]
             )

model_onehot.summary()

Model: "multiclass_classification_model_onehot"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_1 (Flatten)         (None, 4)                 0         
                                                                 
 dense_2 (Dense)             (None, 8)                 40        
                                                                 
 dense_3 (Dense)             (None, 3)                 27        
                                                                 
Total params: 67
Trainable params: 67
Non-trainable params: 0
_________________________________________________________________


In [13]:
layers_integer = [keras.layers.Flatten(input_shape=(4,)),
                  keras.layers.Dense(8, activation=tf.nn.relu),
                  keras.layers.Dense(3, activation=tf.nn.softmax),
                 ]

# Compilación de la red de neuronas, agrupando las diferentas capas de forma secuencial
model_integer = keras.Sequential(layers_integer, name="multiclass_classification_model_integer")

# Configuración del algoritmo de optimización y de la función de loss
# Nota, en este caso vamos a usar Adam en vez de SGD simplemente para demostrar 
# como se le llama en Keras
model_integer.compile(optimizer="adam",
              loss="sparse_categorical_crossentropy",
              metrics=["accuracy"]
             )

model_integer.summary()

Model: "multiclass_classification_model_integer"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_2 (Flatten)         (None, 4)                 0         
                                                                 
 dense_4 (Dense)             (None, 8)                 40        
                                                                 
 dense_5 (Dense)             (None, 3)                 27        
                                                                 
Total params: 67
Trainable params: 67
Non-trainable params: 0
_________________________________________________________________


Con respecto a los parámetros observamos lo siguiente:
* La capa de entrada `flatten_3` no tiene parámetros ya que simplemente nos sirve para introducir información en la red.
* La capa densa `dense_6` está formada por 8 neuronas, cada una de ellas con 4 parámetros de entrada y un parámetro de *bias*, haciendo un total de **40**.
* La capa densa `dense_7` está formada por 3 neuronas con 8 parámetros de entrada y un parámetro de *bias*, haciendo un total de **27**.

### 2.5.3 - Entrenamiento

Una vez compilada nuestra red de neuronas, podemos abordar el proceso de entrenamiento mediante la función `fit`. En este caso, hemos seleccionado las siguientes opciones:
* Número de iteraciones (*epochs*): 125
* Tamaño del batch de entrenamiento: 10 ejemplos
* Tamaño del conjunto de validación: 8% del conjunto de entrenamiento

In [14]:
from keras.callbacks import TensorBoard

# Definición de los callback de TF Board
tensorboard_callback = TensorBoard(log_dir="logs_multiclass_onehot")

# Ejecución del proceso de aprendizaje
model_onehot.fit(
    X_train,
    y_train,
    epochs=125,
    batch_size=10,
    validation_split=0.08,
    callbacks=[tensorboard_callback]
)

Epoch 1/125
13/13 [==============================] - 1s 25ms/step - loss: 2.1820 - accuracy: 0.0000e+00 - val_loss: 2.3769 - val_accuracy: 0.0000e+00
Epoch 2/125
13/13 [==============================] - 0s 6ms/step - loss: 1.9292 - accuracy: 0.0909 - val_loss: 2.0821 - val_accuracy: 0.1818
Epoch 3/125
13/13 [==============================] - 0s 6ms/step - loss: 1.7154 - accuracy: 0.2314 - val_loss: 1.8592 - val_accuracy: 0.1818
Epoch 4/125
13/13 [==============================] - 0s 5ms/step - loss: 1.5647 - accuracy: 0.3140 - val_loss: 1.6739 - val_accuracy: 0.1818
Epoch 5/125
13/13 [==============================] - 0s 5ms/step - loss: 1.4458 - accuracy: 0.3471 - val_loss: 1.5292 - val_accuracy: 0.1818
Epoch 6/125
13/13 [==============================] - 0s 6ms/step - loss: 1.3494 - accuracy: 0.3471 - val_loss: 1.4250 - val_accuracy: 0.1818
Epoch 7/125
13/13 [==============================] - 0s 5ms/step - loss: 1.2774 - accuracy: 0.3388 - val_loss: 1.3335 - val_accuracy: 0.3636
Epoc

In [15]:
from keras.callbacks import TensorBoard

# Definición de los callback de TF Board
tensorboard_callback = TensorBoard(log_dir="logs_multiclass_integer")

# Ejecución del proceso de aprendizaje
model_integer.fit(
    X_train,
    y_train_integer,
    epochs=125,
    batch_size=10,
    validation_split=0.08,
    callbacks=[tensorboard_callback]
)

Epoch 1/125
13/13 [==============================] - 1s 22ms/step - loss: 1.5460 - accuracy: 0.3306 - val_loss: 1.7557 - val_accuracy: 0.3636
Epoch 2/125
13/13 [==============================] - 0s 7ms/step - loss: 1.4141 - accuracy: 0.3306 - val_loss: 1.5975 - val_accuracy: 0.3636
Epoch 3/125
13/13 [==============================] - 0s 6ms/step - loss: 1.3018 - accuracy: 0.3306 - val_loss: 1.4537 - val_accuracy: 0.3636
Epoch 4/125
13/13 [==============================] - 0s 7ms/step - loss: 1.2021 - accuracy: 0.3306 - val_loss: 1.3372 - val_accuracy: 0.3636
Epoch 5/125
13/13 [==============================] - 0s 7ms/step - loss: 1.1177 - accuracy: 0.3884 - val_loss: 1.2155 - val_accuracy: 0.4545
Epoch 6/125
13/13 [==============================] - 0s 5ms/step - loss: 1.0376 - accuracy: 0.5372 - val_loss: 1.1101 - val_accuracy: 0.5455
Epoch 7/125
13/13 [==============================] - 0s 7ms/step - loss: 0.9700 - accuracy: 0.6777 - val_loss: 1.0248 - val_accuracy: 0.5455
Epoch 8/125


In [16]:
# Evaluación de los modelos mediante el conjunto de test
test_loss_onehot, test_acc_onehot = model_onehot.evaluate(X_test, y_test)
test_loss_integer, test_acc_integer = model_integer.evaluate(X_test, y_test_integer)

1/1 [==============================] - 0s 38ms/step - loss: 0.3225 - accuracy: 0.9444


### 2.5.4 - Inferencia

En este caso, en vez de visualizar los resultados obtenidos durante el proceso de entrenamiento, vamos a ver cómo podríamos almacenar el modelo y volver a cargarlo para efecturar predicciones. Así, vamos a almacenar el modelo en el fichero `multiclass_classification_model.h5` y, después, vamos a cargarlo de neuvo para generar las predicciones sobre el conjunto de test comprobando si estas cuadran con el resultado esperado.

In [17]:
model_onehot.save("multiclass_classification_model.h5")

old_model = keras.models.load_model("multiclass_classification_model.h5")

In [18]:
index = 1
predictions_onehot = model_onehot.predict(X_test)
predictions_integer = model_integer.predict(X_test)

print(f"Predicción (one-hot): {predictions_onehot[index]}")
print(f"Predicción (integer): {predictions_integer[index]}\n")
print(f"Real (one-hot): {y_test[index]}")
print(f"Real (integer): {y_test_integer[index]}\n")
print(test_set.iloc[index])

1/1 [==============================] - 0s 67ms/step
Predicción (one-hot): [0.9513948  0.04735072 0.00125445]
Predicción (integer): [9.9021727e-01 9.5272548e-03 2.5544650e-04]

Real (one-hot): [1. 0. 0.]
Real (integer): 0

sepal_length            4.6
sepal_width             3.2
petal_length            1.4
petal_width             0.2
class           Iris-setosa
Name: 47, dtype: object


Como se puede observar, el valor real esperado es la primera clase (Iris-setosa), la cual es modelo predice correctamente con una confianza (probabilidad) de mas del 90%. Esta predicción es correcta.

## 2.6 - Clasificación multietiqueta

A continuación, vamos a describir cómo construir una red de neuronas mediante la utilización de **un conjunto de valores artificiales generados con el objetivo de
explicar este tipo de técnica**. Más adelante volveremos a este tipo de red en el ámbito de clasificación de imágenes, donde los problemas multietiqueta son comunes.

### 2.6.1 - Preparación de datos

La primera fase de todo proceso de ML es la preparación de los datos. En este caso, vamos a construir un conjunto de datos artificial donde contaremos con 16 atributos por cada instancia y 5 posibles etiquetas. Cada una de las instancias de entrenamiento tendrá asignada, al menos, 2 etiquetas:

In [19]:
import tensorflow as tf
import numpy as np

from sklearn.datasets import make_multilabel_classification
from sklearn.model_selection import train_test_split

n_features = 16
n_classes = 5

data_x, data_y = make_multilabel_classification(
    n_samples=1000,
    n_features=n_features,
    n_classes=n_classes,
    n_labels=2,
    random_state=1
)

Una vez realizada la carga de los datos, debemos llevar a cabo una serie de transformaciones y modificaciones en la información del conjunto de datos con el objetivo de adecuarlos y poder construir nuestra red neuronal:

* Codificación numérica de todos los atributos a formato decimal.
* Definición de tensores para representar todas las instancias del dataset.
* Generación de los splits de entrenamiento y test.

In [20]:
# Generación de los conjuntos de entrenamiento y test split_handler (en este caso no vamos a generar de validación aqui, lo haremos mas adelante)
# split proportion: 0.88 train, 0.12 test
x_train, x_test, y_train, y_test = train_test_split(data_x, data_y, test_size=0.10, random_state=0)

# Transformación del conjunto de entrenamiento en tensores
X_train = tf.convert_to_tensor(x_train, np.int32)
y_train = tf.convert_to_tensor(y_train)

# Transformación del conjunto de test en tensores
X_test = tf.convert_to_tensor(x_test, np.int32)
y_test = tf.convert_to_tensor(y_test)

### 2.6.2 - Construcción de la red

Una vez preparados los datos, podemos construir nuestra red neuronal. Para ello, vamos a crear una red formada por tres capas mediante Keras:

* Una capa de de tipo `flatten` que acepte como entrada tensores unidimensionales de dimensión `n_features` (en este caso 16).
* Una capa densa de 32 neuronas con una función de activación de tipo ReLU.
* Una capa densa de `n_classes` (en este caso 5) neuronas con una función de activación de tipo sigmoidea, que nos devolverá la probabilidad de que el ejemplo pertenezca a cada una de las clases.

In [21]:
import tensorflow as tf
from tensorflow import keras

# Creación de las capas que forman la estructura de la red
layers = [keras.layers.Flatten(input_shape=(n_features,)),
                 keras.layers.Dense(32, activation=tf.nn.relu),
                 keras.layers.Dense(n_classes, activation=tf.nn.sigmoid),
                ]

# Compilación de la red de neuronas, agrupando las diferentas capas de forma secuencial
model = keras.Sequential(layers, name="multilabel_classification_model")

# Configuración del algoritmo de optimización y de la función de loss
model.compile(optimizer="adam",
              loss="categorical_crossentropy",
              metrics=["accuracy"]
             )

model.summary()

Model: "multilabel_classification_model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_3 (Flatten)         (None, 16)                0         
                                                                 
 dense_6 (Dense)             (None, 32)                544       
                                                                 
 dense_7 (Dense)             (None, 5)                 165       
                                                                 
Total params: 709
Trainable params: 709
Non-trainable params: 0
_________________________________________________________________


Con respecto a los parámetros observamos lo siguiente:
* La capa de entrada `flatten_4` no tiene parámetros ya que simplemente nos sirve para introducir información en la red.
* La capa densa `dense_8` está formada por 32 neuronas, cada una de ellas con 16 parámetros de entrada y un parámetro de *bias*, haciendo un total de **544**.
* La capa densa `dense_9` está formada por 5 neuronas con 32 parámetros de entrada y un parámetro de *bias*, haciendo un total de **165**.

### 2.6.3 - Entrenamiento

Una vez compilada nuestra red de neuronas, podremos abordar el proceso de entrenamiento mediante la función `fit()`. En este caso, hemos seleccionado las siguientes opciones:
* Numero de epochs: 200
* Tamaño del *batch* de entrenamiento: 250 instancias
* Tamaño del conjunto de validación: 10% del conjunto de entrenamiento

In [22]:
from keras.callbacks import TensorBoard

# Definición de los callback de TF Board
tensorboard_callback = TensorBoard(log_dir="logs_multilabel")

# Ejecución del proceso de aprendizaje
model.fit(
    X_train,
    y_train,
    epochs=200, # Numero de epochs
    batch_size=250, # Tamaño de los batches
    validation_split=0.10, # Tamaño del conjunto de validación
    callbacks=[tensorboard_callback]
)

Epoch 1/200
4/4 [==============================] - 1s 107ms/step - loss: 5.6853 - accuracy: 0.0593 - val_loss: 5.0186 - val_accuracy: 0.0778
Epoch 2/200
4/4 [==============================] - 0s 28ms/step - loss: 5.1625 - accuracy: 0.0716 - val_loss: 4.5595 - val_accuracy: 0.0778
Epoch 3/200
4/4 [==============================] - 0s 18ms/step - loss: 4.7120 - accuracy: 0.0840 - val_loss: 4.1644 - val_accuracy: 0.0778
Epoch 4/200
4/4 [==============================] - 0s 22ms/step - loss: 4.3310 - accuracy: 0.0975 - val_loss: 3.8231 - val_accuracy: 0.0889
Epoch 5/200
4/4 [==============================] - 0s 21ms/step - loss: 4.0015 - accuracy: 0.1136 - val_loss: 3.5338 - val_accuracy: 0.1000
Epoch 6/200
4/4 [==============================] - 0s 21ms/step - loss: 3.7174 - accuracy: 0.1222 - val_loss: 3.2910 - val_accuracy: 0.1333
Epoch 7/200
4/4 [==============================] - 0s 27ms/step - loss: 3.4864 - accuracy: 0.1407 - val_loss: 3.1001 - val_accuracy: 0.1667
Epoch 8/200
4/4 [==

In [23]:
test_loss, test_acc = model.evaluate(X_test, y_test)

4/4 [==============================] - 0s 3ms/step - loss: 2.1761 - accuracy: 0.5200


### 2.6.4 - Inferencia

Simplemente vamos a analizar la salida del modelo con el conjunto de test

In [24]:
predictions = model.predict(X_test)

index = 1
print(f"Predicción: {predictions[index]}")
print(f"Real: {y_test[index]}")

4/4 [==============================] - 0s 2ms/step
Predicción: [0.75951594 0.7472822  0.00602337 0.712368   0.14843097]
Real: [1 1 0 0 0]


## 2.7 - Optimización de parámetros

Los hiperparámetros son un conjunto especial de parámetros que permiten regir el funcionamiento de la fase de entreanamiento. Se utilizan para definir ciertos elementos del algoritmo o de la fase de entranamiento que son independientes del conjunto de datos utilizado. A continuación, se indican algunos de los hiperparámetros más utilizados:

* Tasa de aprendizaje.
* Tamaño de *batch*.
* Número de capas ocultas.
* Número de neuronas de capa oculta.

La **optimización de hiperparámetros** se define como el proceso destinado a hallar la configuración de hiperparámetros que permita generar el modelo con el mejor rendimiento. Para este proceso de búsqueda, realizaremos **múltiples ejecuciones de entrenamiento mediante la modificación conjunta de los diferentes hiperparámetros que se han de optmizar**. En este sentido, **haremos unicamente uso de los conjuntos de entrenamiento y valdiación**, no del de test. Finalmente, cuanto mayor sea el número de hiperparámetros, mayor será el numero de combinaciones a evaluar y con ello el tiempo y coste computacional del proceso de optimización.

Para seleccionar los mejores valores de los hiperparámetros se utilizan diferentes tipos de algoritmos de búsqueda, siendo los más utilizados los que se desriben a continuación:

* **Busqueda aleatoria (Random Search)**, consiste en seleccionar valores para los diferentes hiperparámetros de manera aleatoria dentro del espacio de búsqueda que está formado por los rangos válidos para cada uno de los hiperparámetros.

* **Búsqueda en cuadrícula (Grid Search)**, es un proceso de búsqueda donde los diferentes valores de hiperparámetros se combinan para crear una maya (*grid*) en la cual se incluyen todas las posibles combinaciones de hiperparámetros que serán utilizadas para construir el modelo. En ese caso, el proceso de búsqueda consiste en ir seleccionado posibles configuraciones y probar que resultado ofrece en el momento de construir un modelo de razonamiento

**En resumen, mientras que la búsqueda en cuadrícula analiza todas las combinaciones posibles de hiperparámetros para encontrar el mejor modelo, la búsqueda aleatoria selecciona y prueba una combinaciones aleatorias de hiperparámetros en vez de realizar una búsqueda exhaustiva.**

Si bien éstas son las principales técnicas para la selección de hiperparámetros, en los últimos años se han desarrollado muchas otras, algunas de las cuales se incluyen en los frameworks de Machine Learning y Deep Learning:

* [Búsqueda Bayesiana.](https://towardsdatascience.com/a-conceptual-explanation-of-bayesian-model-based-hyperparameter-optimization-for-machine-learning-b8172278050f)
* [Búsqueda evolutiva mediante la utilización de algoritmos genéticos.](https://medium.com/cindicator/genetic-algorithms-and-hyperparameters-weekend-of-a-data-scientist-8f069669015e)
* [Búsqueda mediante optimización basada en gradientes.](https://en.wikipedia.org/wiki/Hyperparameter_optimization)

Independientemente del tipo de algoritmos de búsqueda que utilicemos, el proceso de optimización de hiperparámetros suele estar formado por cuatro fases:

<img src="images_2/optimizacion_hiperparametros.png" width="800" data-align="center">

* **Definición de hiperparámetros.** Definir los posibles valores que podrán adoptar los diferentes hiperparámetros seleccionados para ser optimizados. Así, por ejemplo, la tasa de aprendizaje solo admitirá valores decimales comprendidos entre 0 y 1, mientras que el número de neuronas de una capa admitirá valores enteros situados entre 1 e infinito.
* **Selección del algoritmo de búsqueda.** Definir el algoritmo que será utilizado para buscar en el espacio de hiperparámetros, así como su configuración. Por ejemplo, el número de iteraciones del proceso de búsqueda.
* **Búsqueda.** Efectuar el proceso de búsqueda. Esta fase consta, a su vez, de tres etapas:
    * Selección. Seleccionar una serie de valores para los hiperparámetros.
    * Entrenamiento. El proceso de entrenamiento con el objetivo de generar un modelo y poder evaluar su calidad mediante el conjunto de validación.
    * Evaluación. Evaluar el modelo generado con el objetivo de crear un ranking en base a la selección de hiperparámetros.
* **Selección de la configuración final.** Evaluar el modelo final con el conjuno de test.

In [25]:
# %pip install keras-tuner --upgrade

### 2.7.1 - Preparación de datos

La primera fase de todo proceso de ML es la preparación de los datos. En este caso, vamos a utilizar **el conjunto de datos sobre pruebas diagnósticas para la detección del cáncer de mama**.

In [26]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf

data = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data", header=None)
data.columns = [
    "Sample_code_number",
    "Clump_Thickness",
    "Uniformity_of_Cell_Size",
    "Uniformity_of_Cell_Shape",
    "Marginal_Adhesion",
    "Single_Epithelial_Cell_Size",
    "Bare_Nuclei",
    "Bland_Chromatin",
    "Normal_Nucleoli",
    "Mitoses",
    "Class"
]

Una vez realizada la carga de los datos, debemos llevar a cabo una serie de transformaciones y modificaciones en la información del conjunto de datos con el objetivo de adecuarlos y poder construir nuestra red de neuronas:

In [27]:
# Eliminación de la columna que contiene el id de las muestras 
data.drop("Sample_code_number", axis="columns", inplace=True)

# Modificación de los valores ? que existen en la columna Bare Nuclei (en este caso por el valor -1)
data["Bare_Nuclei"] = data['Bare_Nuclei'].replace('?', '-1')

# Modificación de las etiquetas para que los valores sean 0 y 1 
data["Class"] = data["Class"].map(lambda x: 1 if x == 4 else 0)

# Generación de los conjuntos de entrenamiento y test split_handler (en este caso no vamos a generar de validación aqui, lo haremos mas adelante)
# split proportion: 0.88 train, 0.12 test
train_set, test_set = train_test_split(data, test_size=0.12, random_state=0, stratify=data["Class"])

# Transformación del conjunto de entrenamiento en tensores 
X_train = tf.convert_to_tensor (train_set.iloc[:, :9], np.int32) 
y_train = tf.convert_to_tensor (train_set.iloc[:, 9:], np.int8)

# Transformación del conjunto de test en tensores 
X_test = tf.convert_to_tensor (test_set.iloc[:, :9], np.float32)
y_test = tf.convert_to_tensor (test_set.iloc[:, 9:], np.int8)

### 2.7.2 - Definición de hiperparámetros

Es necesario definir los diferentes hiperparámetros que se han de optimizar. En este sentido, **vamos a utilizar el sistema de optimización de Keras**, que nos permite llevar a cabo dicho definición de manera muy sencilla. Para desarrollar nuestro ejemplo, vamos a optimizar dos parámetros de la capa densa de nuestra red de neuronas:
* **Número de neuronas de la capa oculta**.
    * Valor mínimo (`min_value`): 9.
    * Valor máximo (`max_value`): 27.
    * Paso (`step`): variación que se producirá sobre el parámetro en las sucesivas iteraciones del algoritmo de búsqueda y que poseerá un valor de 3.
    * Valor inicial (`default`): 12.
* **Función de activación**.  El segundo parámetro que se va a optimizar será la función de activación de la capa oculta. Para ello **permitiremos que esta pruebe dos posibles funciones de activación: ReLU y tangente hiperbólica**.

In [28]:
import tensorflow as tf

from tensorflow import keras
from keras.callbacks import TensorBoard
from keras_tuner import HyperModel

class FFNNHyperModel(HyperModel):
    
    def build(self, hp):
        
        hp_units = hp.Int(
            "dense_1_units",
            min_value=9,
            max_value=27,
            step=3,
            default=12
        )
        
        hp_function = hp.Choice(
            "dense_1_activation",
            values=["relu", "tanh"],
            default="relu"
        )
        
        layers = [
            keras.layers.Flatten(name="input", input_shape=(9,)),
            keras.layers.Dense(
                name="hidden",
                units=hp_units,
                activation=hp_function
            ),
            keras.layers.Dense(
                name="output",
                units=1,
                activation=tf.nn.sigmoid
            ),
        ]
        
        model = keras.Sequential(layers, name="binary_classification")
        
        model.compile(
            optimizer="sgd",
            loss="binary_crossentropy",
            metrics=["accuracy"]
        )
        
        model.summary()
        
        return model

### 2.7.3 - Definición del proceso de búsqueda

En este caso, vamos a definir:
* **El algoritmo de búsqueda**, en este caso haremos uso del más simple, es decir, la búsqueda aleatoria. 
    * Definiremos junto a ella una semilla (*seed*) para poder reproducir el proceso de búsqueda de nuevo
* **Número de intentos que se realizarán**.
* **Numero de ejecuciones en cada intento**. Número de modelos que se generarán para conjunto de hiperparámetros. Podemos considerar un número más o menos alto para intentar reducir la variabilidad de aprendizaje, principalmente para evitar descartar modelos que "tuvieron mala suerte" durante el aprendizaje.

In [29]:
from keras_tuner.tuners import RandomSearch

MAX_TRIALS = 25
EXECUTIONS_PER_TRIAL = 2
SEED = 0

hypermodel = FFNNHyperModel()

tuner = RandomSearch(
    hypermodel,
    objective="val_accuracy",
    seed=SEED,
    max_trials=MAX_TRIALS,
    executions_per_trial=EXECUTIONS_PER_TRIAL,
    directory="random_search",
    project_name="tokio_tema_3"
)

tuner.search_space_summary()

INFO:tensorflow:Reloading Tuner from random_search\tokio_tema_3\tuner0.json
Search space summary
Default search space size: 2
dense_1_units (Int)
{'default': 12, 'conditions': [], 'min_value': 9, 'max_value': 27, 'step': 3, 'sampling': 'linear'}
dense_1_activation (Choice)
{'default': 'relu', 'conditions': [], 'values': ['relu', 'tanh'], 'ordered': False}


Aqui podemos ver un resumen tanto del modelo inicial como del proceso de búsqueda que se va a ejecutar.

### 2.7.4 - Ejecución del proceso de búsqueda

En este caso debemos definir los parámetros de configuracion relativos a la ejecución:
* El número de epochs para cada ejecución (vamos a usar el valor que vimos en la sección 2.4)
* El tamaño del conjunto de validación (vamos a usar el valor que vimos en la sección 2.4)

----

**Nota:** Simplemente como reflexión, podriamos también realizar un proceso de optimización (de más alto nivel) sobre los hiperparámetros del propio proceso de optimización. En ese caso estariamos en el campo de **meta-aprendizaje**. Una heurística para seleccionar valores aqui es tener una idea de experimentos pasados, ya sean propios o de la literatura.

----

In [30]:
N_EPOCH_SEARCH = 125

tuner.search(
    X_train,
    y_train,
    epochs=N_EPOCH_SEARCH,
    validation_split=0.12
)

Trial 14 Complete [00h 00m 18s]
val_accuracy: 0.9391891658306122

Best val_accuracy So Far: 0.9391891658306122
Total elapsed time: 00h 03m 25s
INFO:tensorflow:Oracle triggered exit


### 2.7.5 - Selección del la configuración final

In [31]:
tuner.results_summary()

best_model = tuner.get_best_models(num_models=1)[0]

loss, accuracy = best_model.evaluate(X_test, y_test)

print(loss)
print(accuracy)

Results summary
Results in random_search\tokio_tema_3
Showing 10 best trials
Trial summary
Hyperparameters:
dense_1_units: 15
dense_1_activation: tanh
Score: 0.9391891658306122
Trial summary
Hyperparameters:
dense_1_units: 27
dense_1_activation: tanh
Score: 0.9391891658306122
Trial summary
Hyperparameters:
dense_1_units: 21
dense_1_activation: relu
Score: 0.9324324131011963
Trial summary
Hyperparameters:
dense_1_units: 24
dense_1_activation: tanh
Score: 0.9256756603717804
Trial summary
Hyperparameters:
dense_1_units: 9
dense_1_activation: relu
Score: 0.9256756603717804
Trial summary
Hyperparameters:
dense_1_units: 18
dense_1_activation: relu
Score: 0.9256756603717804
Trial summary
Hyperparameters:
dense_1_units: 24
dense_1_activation: relu
Score: 0.9256756603717804
Trial summary
Hyperparameters:
dense_1_units: 27
dense_1_activation: relu
Score: 0.9256756603717804
Trial summary
Hyperparameters:
dense_1_units: 18
dense_1_activation: tanh
Score: 0.9189189076423645
Trial summary
Hyperparam